driver-core/src/main/com/mongodb/internal/connection/DefaultConnectionPool.java

```diff
@@ -1170,55 +1154,170 @@ int getNumPinnedToTransaction() {
         }
     }
 
+    /**
+     * This class maintains threads needed to perform {@link #getAsync(SingleResultCallback)}.
+     */
     @ThreadSafe
-    private static class Workers implements AutoCloseable {
-        private volatile ExecutorService getter;
-        private volatile ExecutorService opener;
+    private static class AsyncWorkManager implements AutoCloseable {
+        private volatile State state;
+        private volatile BlockingQueue<Task> tasks;
         private final Lock lock;
+        @Nullable
+        private ExecutorService worker;
 
-        Workers() {
+        AsyncWorkManager() {
+            state = State.NEW;
+            tasks = new LinkedBlockingQueue<>();
             lock = new StampedLock().asWriteLock();
         }
 
-        Executor getter() {
-            if (getter == null) {
-                lock.lock();
-                try {
-                    if (getter == null) {
-                        getter = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
-                    }
-                } finally {
-                    lock.unlock();
+        void enqueue(final Task task) {
+            lock.lock();
+            try {
+                if (initUnlessClosed()) {
+                    tasks.add(task);
+                    return;
                 }
+            } finally {
+                lock.unlock();
             }
-            return getter;
+            task.failAsClosed();
         }
 
-        Executor opener() {
-            if (opener == null) {
-                lock.lock();
-                try {
-                    if (opener == null) {
-                        opener = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncOpener"));
-                    }
-                } finally {
-                    lock.unlock();
-                }
+        /**
+         * @return {@code false} iff the {@link #state} is {@link State#CLOSED}.
+         */
+        private boolean initUnlessClosed() {
+            boolean result = true;
+            if (state == State.NEW) {
+                worker = Executors.newSingleThreadExecutor(new DaemonThreadFactory("AsyncGetter"));
+                assertNotNull(worker).submit(() -> runAndLogUncaught(this::workerRun));
+                state = State.INITIALIZED;
+            } else if (state == State.CLOSED) {
+                result = false;
             }
-            return opener;
+            return result;
         }
 
+        /**
+         * {@linkplain Thread#interrupt() Interrupts} all workers and causes queued tasks to
+         * {@linkplain Task#failAsClosed() fail} asynchronously.
+         */
         @Override
+        @SuppressWarnings("try")
         public void close() {
+            lock.lock();
             try {
-                if (getter != null) {
-                    getter.shutdownNow();
+                if (state != State.CLOSED) {
```

Comment:
I would find:

                if (state != State.CLOSED) {
                    state = State.CLOSED;
                    if (worker != null) {
                        worker.shutdownNow();
                    }
                }

easier to understand.  Is there an important functional difference that I'm not seeing?

Reply:
Thanks, good point. This code was written while there were two workers and a timeouter. Using try-with-resources made sense at that point, but with a single worker we are better off without it.

Done.